### ISSUE TITLE
Pytorch学习

### ISSUE THREAD

**Issue by LegendBC** [LegendBC/Interest-Explore#2]

**Comment by LegendBC** [comment]
Pytorch hook使用

https://youtu.be/syLFCVYua6Q

PyTorch 15.hook技术 - Pytorch日积月累 - 知乎专栏 https://zhuanlan.zhihu.com/p/152314451?utm_source=ZHShareTargetIDMore&utm_medium=social&utm_oi=785040813519433728

**Comment by LegendBC** [comment]
关于多卡训练，以及训练加速的方法：
https://glassboxmedicine.com/2020/03/04/multi-gpu-training-in-pytorch-data-and-model-parallelism/
1. model parallel
2. data parallel
3. multi workers to load data
4. distributeddataparallel, 多进程； dataparall， 多线程

**Comment by LegendBC** [comment]
关于DDP的使用：
1. get local rank and set gpu device
2. init process group
3. get model and mov model to corresponding device
4. get optimizer
5. sync batchnorm
6. sample dataloader
7. set sampler epoch
8. destroy process group at the end of training
